[SSP-2007] entity changes log (#2980)
shopsys/shopsys: docs/cookbook/log-entity-changes.md

# Log entity changes

> The Shopsys platform allows you to log changes to entities.

Reviewer: ```suggestion
The Shopsys platform allows you to log changes to entities. Thanks to the mechanism, you can answer the questions like "Who and when changed the order status?", "Why is the product price changed?", etc.
```

Author: ok

The logging system works to capture changes in the doctrine unit of work before they are flushed.

> After setting logging we used php attributes above class, property and method. (https://www.php.net/manual/en/language.attributes.overview.php)

Reviewer 2: ```suggestion
We used PHP attributes above class, property, and method for setting logging. (https://www.php.net/manual/en/language.attributes.overview.php)
```

Author: fixed

> As an implemented sample, you can study the settings on Order, OrderItem, OrderStatus and Country entities. The principle of use will be described on these examples.

Reviewer 2: It would be nice to have each sentence on a separate line. It improves the readability of a markdown file (and diffs are nicer), while it will still be rendered as a single paragraph.

I don't insist on the change in this already written article (although I would not be against it 😄 ). 
Just for the future.

Author: sure

The attributes `Loggable` and `LoggableChild` are used to mark the entity to be logged. Both of these attributes have logging strategy settings available. If I want to log all properties in the base, I use the `Loggable(Loggable::STRATEGY_INCLUDE_ALL)` strategy. If I would like to not log certain properties, I can mark them with the `ExcludeLog` attribute. Conversely, if I want to log only a few properties from an entity, it would be better to use the `Loggable(Loggable::STRATEGY_EXCLUDE_ALL)` strategy and then mark which properties I want to log using the `Log` attribute.
!!! danger

>     **Now is possible log only entity with primary key `id` and method `getId()`**

Reviewer 2: ```suggestion
    **It is possible to log only entities with primary key `id` and method `getId()`**
```

Author: sure

 * @ORM\Entity
 */
#[LoggableChild(Loggable::STRATEGY_INCLUDE_ALL)]
class OrderItem
{
    ...

    /**
     * @var \Shopsys\FrameworkBundle\Model\Order\Order
     * @ORM\ManyToOne(targetEntity="Shopsys\FrameworkBundle\Model\Order\Order", inversedBy="items")
     * @ORM\JoinColumn(name="order_id", referencedColumnName="id", nullable=false, onDelete="CASCADE")
     */
    #[LoggableParentProperty]
    protected $order;
    ...
}
```

## Option to name the logged record

The administrator wants to see a human-readable record for the logged record. This is what the `EntityLogIdentify` attribute is used for. This attribute is set on the method that should return the name of the entity. For the OrderItem entity, this is the `getName()` method. Some entities are translatable and you need to mark them like this: `EntityLogIdentify(EntityLogIdentify::IS_LOCALIZED)`. In the background, the administration locale is inserted when such a method is called.

```php
#[LoggableChild(Loggable::STRATEGY_INCLUDE_ALL)]
class OrderItem
{
    ...

    /**
     * @return string
     */
    #[EntityLogIdentify]
    public function getName()
    {
        return $this->name;
    }

    ...
}
```

> There are properties in the Order entity that are not only of scalar data type. For example, the status property is of data type OrderStatus or deliveryCountry is of data type Country (another entity). These entities themselves do not need to be logged in, but we want to see the human name on the order in the event of a status change. The `EntityLogIdentify` attribute is again used for this naming. So if I want to see in the order log that the status has changed from New to In Progress, I need to mark the getName function on the OrderStatus entity.

Reviewer 2: ```suggestion
There are properties in the Order entity that are not a simple scalar data type. For example, the status property is of data type OrderStatus or deliveryCountry is of data type Country (another entity). These entities themselves do not need to be logged, but we want to see the human name on the order in the event of a status change. The `EntityLogIdentify` attribute is again used for this naming. So if I want to see in the order log that the status has changed from New to In Progress, I need to mark the getName function on the OrderStatus entity.
```

Author: sure

> There are properties in the Order entity that are not only of scalar data type. For example, the status property is of data type OrderStatus or deliveryCountry is of data type Country (another entity). These entities themselves do not need to be logged in, but we want to see the human name on the order in the event of a status change. The `EntityLogIdentify` attribute is again used for this naming. So if I want to see in the order log that the status has changed from New to In Progress, I need to mark the getName function on the OrderStatus entity.

Reviewer: ```suggestion
There are properties in the Order entity that are not only of scalar data type. For example, the status property is of data type OrderStatus or deliveryCountry is of data type Country (another entity). These entities themselves do not need to be logged, but we want to see the human name on the order in the event of a status change. The `EntityLogIdentify` attribute is again used for this naming. So if I want to see in the order log that the status has changed from "New" to "In Progress", I need to mark the getName function on the OrderStatus entity.
````

> ```php

Reviewer: the example does not match with the explanation above - the `OrderStatus` is mentioned there, so IMHO the code should show the `OrderStatus` entity, not `Country`

Author: thx. fixed
